tests/installation/select_patterns_and_packages.pm

```diff
@@ -155,113 +141,127 @@ sub package_action {
-                        $current_pattern      = $p;
-                        # add pattern to the set of detected patterns
-                        $processed_patterns{$p} = undef;
-                        record_info($current_pattern, $needs_to_be_selected);
-                    }
-                }
-            }
-            $needs_to_be_selected = 1 if ($patterns{all});
+sub select_specific_patterns_by_iteration {
+    my %patterns;              # set of patterns to be processed
+    my %processed_patterns;    # set of the processed patterns
 
-            my $selected = check_screen([qw(current-pattern-selected on-category)], 0);
+    # fill with variable values
+    @patterns{split(/,/, get_var('PATTERNS'))} = ();
 
-            if ($selected && $selected->{needle}->has_tag('on-category')) {
-                movedownelseend;
-                check12qtbug;
-                next;
-            }
-            if ($needs_to_be_selected && !$selected) {
-                wait_screen_change {
-                    send_key ' ';
-                    record_info("select", "");
-                };
-                assert_screen 'current-pattern-selected', 5;
-            }
-            # stick to the default patterns. Check if at least 1 dep. issue was displayed
-            $dep_issue = $self->workaround_dependency_issues || $dep_issue;
+    my $counter = 80;
+    # delete special 'all' and 'default' keys from the check
+    delete $patterns{default};
+    delete $patterns{all};
+    while (1) {
+        die "looping for too long" unless ($counter--);
+        my $needs_to_be_selected;
+        my $ret = check_screen('on-pattern', 1);
+        # this variable will only be updated when the pattern list entry is under the cursor
+        # and the pattern is in the PATTERNS variable and there is a needle for that entry
+        my $current_pattern = 'UNKNOWN_PATTERN';
 
-            if (get_var('PATTERNS', '') =~ /default/ && !(get_var('PATTERNS', '') =~ /$current_pattern/)) {
-                $needs_to_be_selected = $selected;
-                record_info("keep default", "");
-            }
-            if (!$needs_to_be_selected && $selected) {
-                send_key ' ';
-                record_info("unselect", "");
-                assert_screen [qw(current-pattern-unselected current-pattern-autoselected)], 8;
+        if ($ret) {    # unneedled pattern
+            for my $p (keys %patterns) {
+                my $sel = 1;
+                if ($p =~ /^-/) {
+                    # this pattern shall be deselected as indicated by '-' prefix
+                    $sel = 0;
+                    $p =~ s/^-//;
+                }
+                if (match_has_tag("pattern-$p")) {
+                    $needs_to_be_selected = $sel;
+                    $current_pattern      = $p;
+                    # add pattern to the set of detected patterns
+                    $processed_patterns{$p} = undef;
+                    delete $patterns{$p};
```

Comment:
This won't work in case of `-pattern` case, as you will try to remove non-existing key here, we should take the key value before the modification on L#199.

Reply:
I missed that, right, good that you saw it. Stored variable untouched.